system/Database/BaseBuilder.php

```diff
@@ -1719,6 +1726,252 @@ public function getWhere($where = null, ?int $limit = null, ?int $offset = 0, bo
+        }
+
+        for ($i = 0, $total = count($set); $i < $total; $i += $batchSize) {
+            if ($hasQBSet) {
+                $QBSet = array_slice($this->QBSet, $i, $batchSize);
+            } else {
+                $this->setBatch(array_slice($set, $i, $batchSize), '', $escape);
+                $QBSet = $this->QBSet;
+            }
+            $sql = $this->{$renderMethod}($table, $this->QBKeys, $QBSet);
+
+            if ($this->testMode) {
+                $savedSQL[] = $sql;
+            } else {
+                $this->db->query($sql, null, false);
+                $affectedRows += $this->db->affectedRows();
+            }
+
+            if (! $hasQBSet) {
+                $this->resetRun([
+                    'QBSet'  => [],
+                    'QBKeys' => [],
+                ]);
+            }
+        }
+
+        $this->resetWrite();
+
+        return $this->testMode ? $savedSQL : $affectedRows;
+    }
+
+    /**
+     * Allows key/value pairs to be set for batch inserts/upserts
+     *
+     * @param mixed $key
+     *
+     * @return $this|null
+     */
+    public function setBatch($key, string $value = '', ?bool $escape = null)
+    {
+        $key = $this->batchObjectToArray($key);
+
+        if (! is_array($key)) {
+            $key = [$key => $value];
+        }
+
+        $escape = is_bool($escape) ? $escape : $this->db->protectIdentifiers;
+
+        $keys = array_keys($this->objectToArray(current($key)));
+        sort($keys);
+
+        foreach ($key as $row) {
+            $row = $this->objectToArray($row);
+            if (array_diff($keys, array_keys($row)) !== [] || array_diff(array_keys($row), $keys) !== []) {
+                // batch function above returns an error on an empty array
+                $this->QBSet[] = [];
+
+                return null;
+            }
+
+            ksort($row); // puts $row in the same order as our keys
+
+            $clean = [];
+
+            foreach ($row as $rowValue) {
+                $clean[] = $escape ? $this->db->escape($rowValue) : $rowValue;
+            }
+
+            $row = $clean;
+
+            $this->QBSet[] = $row;
+        }
+
+        foreach ($keys as $k) {
+            $this->QBKeys[] = $this->db->protectIdentifiers($k, false);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Compiles an upsert query and returns the sql
+     *
+     * @throws DatabaseException
+     *
+     * @return bool|string
+     */
+    public function getCompiledUpsert(bool $reset = true)
+    {
+        $currentTestMode = $this->testMode;
+
+        $this->testMode = true;
+
+        $sql = implode(";\n", $this->upsert());
+
+        $this->testMode = $currentTestMode;
+
+        // this doesn't work with current implimentation - is cleared in upsert method
+        if ($reset === true) {
+            $this->resetWrite();
+        }
+
+        return $this->compileFinalQuery($sql);
+    }
+
+    /**
+     * Converts call to batchUpsert
```

Comment:
What do you mean?

Reply:
Meaning it changes the data set from array to an array of arrays. Then it handles things as a batchUpsert of one row. 
$set = [$set};

Reply:
insert() returns Query Object
insertBatch() returns Number of rows inserted or FALSE on failure, SQL array when testMode
upsert() and upsertBatch() return Number of rows inserted or FALSE on failure, SQL array when testMode

Reply:
> insert() returns Query Object

No, `insert()` returns `bool`.

Reply:
> upsert() and upsertBatch() return Number of rows inserted or FALSE on failure, SQL array when testMode

Number of rows inserted? Not include updated rows?

Reply:
This gets messy with upserts. The data returned at least with myself is almost useless.

Per mysql:
>With ON DUPLICATE KEY UPDATE, the affected-rows value per row is 1 if the row is inserted as a new row and 2 if an existing row is updated.

And if not inserted or updated then 0.

Reply:
That is the return value of `upsertBatch()` with MySQL is not *Number of rows inserted*.

> With ON DUPLICATE KEY UPDATE, the affected-rows value per row is 1 if the row is inserted as a new row, 2 if an existing row is updated, and 0 if an existing row is set to its current values. If you specify the CLIENT_FOUND_ROWS flag to the [mysql_real_connect()](https://dev.mysql.com/doc/c-api/8.0/en/mysql-real-connect.html) C API function when connecting to [mysqld](https://dev.mysql.com/doc/refman/8.0/en/mysqld.html), the affected-rows value is 1 (not 0) if an existing row is set to its current values.
https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html

Reply:
I added to the tests to assert return values for the different DBMS. I included a default value which would show the ideal results - though none of them actually return this. I highlighted the unique behaviors of each DBMS and how they differ from the ideal.

Reply:
I fixed _upsertBatch() where it throws an error if no constraints are found. I added  if ($this->db->DBDebug)  throw error else return false. This way if this error is thrown the executeBatch() can return false on error.

Reply:
added testNoConstraintFound(). Can't test for return false without a setter.
$this->db->DBDebug = false; (protected property) 
Used return ''; // @codeCoverageIgnore